vortex-serde/src/layouts/read/batch.rs

```diff
@@ -2,47 +2,110 @@ use std::mem;
 use std::sync::Arc;
 
 use vortex::array::StructArray;
+use vortex::validity::Validity;
 use vortex::{Array, IntoArray};
-use vortex_error::{vortex_err, VortexResult};
+use vortex_dtype::FieldNames;
+use vortex_error::{vortex_bail, vortex_err, VortexExpect, VortexResult};
+use vortex_expr::VortexExpr;
 
+use crate::layouts::read::selection::RowSelector;
 use crate::layouts::read::{LayoutReader, ReadResult};
+use crate::layouts::{Message, RangeResult};
 
 #[derive(Debug)]
-pub struct BatchReader {
-    names: Arc<[Arc<str>]>,
+pub struct ColumnBatchReader {
+    names: FieldNames,
     children: Vec<Box<dyn LayoutReader>>,
+    read_ranges: Vec<Option<RowSelector>>,
     arrays: Vec<Option<Array>>,
 }
 
-impl BatchReader {
-    pub fn new(names: Arc<[Arc<str>]>, children: Vec<Box<dyn LayoutReader>>) -> Self {
+impl ColumnBatchReader {
```

Comment:
There's duplication here because we sometimes return ranges and sometimes return arrays which is... dumb